extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -679,87 +694,204 @@ public String getPosition(StreamPartition<String> partition)
+          // that we never find the first record in the shard if we use a limit of 1.
+          recordsResult = kinesis.getRecords(
+              new GetRecordsRequest().withShardIterator(shardIterator).withLimit(FETCH_SEQUENCE_NUMBER_RECORD_COUNT)
+          );
+        }
+        catch (ProvisionedThroughputExceededException e) {
+          log.warn(
+              e,
+              "encountered ProvisionedThroughputExceededException while fetching records, this means "
+              + "that the request rate for the stream is too high, or the requested data is too large for "
+              + "the available throughput. Reduce the frequency or size of your requests. Consider increasing "
+              + "the number of shards to increase throughput."
+          );
+          try {
+            Thread.sleep(PROVISIONED_THROUGHPUT_EXCEEDED_BACKOFF_MS);
+            continue;
+          }
+          catch (InterruptedException e1) {
+            log.warn(e1, "Thread interrupted!");
+            Thread.currentThread().interrupt();
+            break;
+          }
+        }
 
-    resource.shardIterator = wrapExceptions(() -> kinesis.getShardIterator(
-        partition.getStream(),
-        partition.getPartitionId(),
-        iteratorEnum.toString(),
-        sequenceNumber
-    ).getShardIterator());
+        List<Record> records = recordsResult.getRecords();
+
+        if (!records.isEmpty()) {
+          return records.get(0).getSequenceNumber();
+        }
+
+        shardIterator = recordsResult.getNextShardIterator();
+      }
+
+      if (shardIterator == null) {
+        log.info("Partition[%s] returned a null shard iterator, is the shard closed?", partition.getPartitionId());
+        return KinesisSequenceNumber.END_OF_SHARD_MARKER;
+      }
+
+
+      // if we reach here, it usually means either the shard has no more records, or records have not been
+      // added to this shard
+      log.warn(
+          "timed out while trying to fetch position for shard[%s], millisBehindLatest is [%s], likely no more records in shard",
+          partition.getPartitionId(),
+          recordsResult != null ? recordsResult.getMillisBehindLatest() : "UNKNOWN"
+      );
+      return null;
+    });
   }
 
-  private void filterBufferAndResetFetchRunnable(Set<StreamPartition<String>> partitions) throws InterruptedException
+  /**
+   * Given a {@link StreamPartition} and an offset, create a 'shard iterator' for the offset and fetch a single record
+   * in order to get the lag: {@link GetRecordsResult#getMillisBehindLatest()}. This method is thread safe as it does
+   * not depend on the internal state of the supplier (it doesn't use the {@link PartitionResource} which have been
+   * assigned to the supplier), and the Kinesis client is thread safe.
+   */
+  private Long getPartitionResourcesTimeLag(StreamPartition<String> partition, String offset)
```

Comment:
Suggest renaming this to not have `Resources` in the name since it doesn't use PartitionResource

Reply:
Woops, this was a rename done with intellij that caught this method in the crossfire, I only meant to rename `getPartitionResourcesTimeLag()` since it does actually get the lag from the `PartitionResource` collection. Fixed.